### PR TITLE
Added Base Module symbols highlighting support

### PR DESCRIPTION
**Pull request by ghyatzo**
Hello again,

Another PR with a big improvment and some minor ones:

Big one: 
- the syntax now recognises all built-in classes, typealiases, and methods and offers some  rough logic i.e.: you can't extend a non `open` class, so writing `class Bird extends Int { ... } ` will mark `Int` as invalid since `Int` is not `open`.
Or also trying to instantiate and `extern` class will also lead to complaints `a = new String {...} //illegal`.
- Added support to constants and methods from the `math` module
- All built-in classes and types are scoped as `storage.type.class.pkl` to differentiate them from custom identifiers that are scoped as `entity.name.class.pkl`.
 - typealiases scoped as `storage.type.enum.pkl`
- built-in methods are scoped as `support.function.pkl`

One big thing missing, since I am not completely sure how to proceed, is trying to scope built-in classes properties, for example: the property of `Int`s `isEven`: `3.isEven` or the property `output` of the module class? should these be scoped as "special" or just leave them unscoped?

Also, I am really not sure how to scope the built-in types and classes, as mentioned I have used `storage.type.class.pkl` so far, but ideally and following the indications of the scope naming page as well as looking at other languages, they could potentially be scoped as `support.type.pkl` instead.
I am more inclined to use `support.type.pkl` since it is more in line with other syntaxes, the scope naming convention and the fact that we are using `support.function.pkl` for the methods of the base module

Other minor changes:
- better function and lambdas parsing
- better typealias parsing
- addressed #3, sorry my syntax doesn't highlight them so didn't realise! now they are scoped as `meta.{property,object,entry,name}.pkl` where `name` is the catchall value.
- some fixes overall and added tests in the syntax file

### CODE REVIEW

**Comment by serjan-nasredin** [comment]
Hi @ghyatzo ✋, thanks for the PR!

**Comment by serjan-nasredin** [comment]
I don't understand why the coloring has changed?
image
Maybe it would be better this way? if these are keywords (`keyword`/`support.function`)

https://github.com/serjan-nasredin/pkl.tmbundle/assets/67647968/b1b93a84-9003-492c-88df-72ff0023df41

image
is it my imagination or is "name" glowing white?
image
Wouldn't it be better to capture the `\` symbol as well? I would also like to note that coloring them in the color "keyword.*.pkl" would be the best idea.
```
\( 1 )
^--- keyword.pkl
   ^--- constant.pkl
      ^--- keyword.pkl
```
image

**Comment by ghyatzo** [comment]
> Maybe it would be better this way? if these are keywords (keyword/support.function)

I am afraid you will need to be a bit more specific regarding which coloring you are talking about. 
If you refer to the `yamlRender` text, that is because I am scoping it as `storage.type.pkl` as I am doing with all built in classes from the base module. But I am planning on using the scope `support.type.pkl` to be more consistent with other syntaxes. See discussion in the PR. If `new` has the same colour as that `yamlRenderer` then it means that your color scheme is giving `storage.type.pkl` and `keyword.pkl` the same colour.
EDIT: I just saw the video about `print`, not sure what you are referring to.

I don't know why that `name` is glowing white, the syntax shouldn't give it nothing other than meta.name.plk, which line is it on?

Well, the `\` is the escape character, I am unsure about giving it a `keyword` scope since it is not really a keyword and in some color schemes could show up as unnecessarily highlighted.
I agree though that it should be of a single color with the parentheses, a good middle ground could be scoping also the parentheses as escape characters or the `\`.

**Comment by serjan-nasredin** [comment]
Hi! Yeah, `YamlRenderer` is colored in the color `keyword.pkl`, but I meant that after the keyword `new` the coloring should go as before. I want to add, `JsonRenderer` and `YamlRenderer`, whatever colors they were painted in other places, when writing the keyword `new` should display the old coloring. As shown in the video example, the color of the word `print` with `keyword` changes when the character `(` is added before the word.

I think if you wrap the characters `\(...)` in `keyword.pkl`, then the numbers will be able to be displayed correctly:
image

It was the `base` variable that confused me. The expression `base = new Base {}` looks terrible.
image

In conclusion, I would like to add that the text wrapped in the characters <code>`</code> should be colored somehow, because they are not visible to the eye, which means that debugging the code will be much more difficult. If you find information that this is a text/variable, then color it in the appropriate color.

**Comment by ghyatzo** [comment]
Hi, before I reply to your issues I would like to make a comment regarding color and scoping.

The role of the syntax is not to color things, the role of the syntax is to scope things, give them a tag or a name. which then can be used by the color-scheme to give colors to different things based on their scope.
In many of your comments I see these two concepts mixed together, but I would like to stress that we are not trying to make a syntax that works for a single specific color-scheme but (possibly) all of them, that means thinking about how a concept in the language is best described in terms of scoping (i.e.: `entity.name.class.pkl` vs `storage.type.pkl`). Sometimes even if the scoping is correct it might be the colorscheme that made some peculiar choices. 
Please understand I am not trying to say that your color-scheme is wrong! just that we should separate the two things and talk about scopes instead of colours :)

Having said that: 
- regarding the `YamlRenderer` after the new, it is colored the same as the `keyword.*.pkl` scope because your color scheme colors `storage.type.*` the same as `keyword.*` (check the scoping, but I am 99% sure the scoping is correct there) so that is something not really relative to the syntax matching per se. built-in types and classes (such as `yamlRenderer`) are scoped differently from user-defined entities, I changed it to `support.type.pkl` for the built-ins, maybe your color-scheme picks up on that difference.
- I am not sure what is the error you have with `\(...)` on my side it is scoping things correctly inside the parens. I agree that the `\` and the `(` should have the same color so I changed the scoping of `\` to match `(` when we do string interpolation and left it as it is when it is a normal escape character.
- yes `base = new Base {}` was confusing because there was an error in the `extends` parses that would not pop off well. fixed.
- Your point on quoted variables is completely valid, I added a `variable.other.pkl` scope to the quoted variables, to let a color-scheme pick that up, and highlight it somehow.

EDIT: regarding `print` example, matching function calls is rather hard, and still needs to be done, left it as last since I believe it was lower priority over other stuff.

**Comment by serjan-nasredin** [comment]
LGTM

The characters are displayed incorrectly:
image
image

**Comment by ghyatzo** [comment]
Nice catch: should be fixed now, also the quoted variable.

regarding the conflict: thanks, missed a cleaner way to write it! that wasn't matching against normal read and import but fixed it.

Let me know of anything else!

**Comment by serjan-nasredin** [comment]
LGTM

**Comment by ghyatzo** [comment]
remember to make a new release 🥳

**Comment by serjan-nasredin** [comment]
Thanks again for the PR! ❤️
> remember to make a new release 🥳

Sure :octocat: !
